### PR TITLE
Upgrade clang version to 14.0.3 in CI pipeline.

### PR DESCRIPTION
**Pull request by usamasaqib**
### What does this PR do?
This PR upgrades the clang and llvm version in the CI pipeline from 11.0.1 to 14.0.3.
The table below compares the sizes (in MB) of the binaries.

||clang arm64 | clang amd64 | llc arm64 | llc amd64 |
|--|---|----|----|---|
| **11.0.1** | 65 | 62 | 22 | 21 |
| **14.0.3** | 77 | 73 | 30 | 29 |

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] At least one `team/..` label has been applied, indicating the team(s) that should QA this change.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

File: .gitlab/deps_build.yml [hunk 1]
```diff
@@ -19,14 +19,18 @@
...
+    - wget ${GENERAL_ARTIFACTS_CACHE_BUCKET_URL}/llvm/clang-14.0.3.src.tar.xz -O clang.src.tar.xz
+    - wget ${GENERAL_ARTIFACTS_CACHE_BUCKET_URL}/llvm/llvm-14.0.3.src.tar.xz -O llvm.src.tar.xz
     - echo -e "${LLVM_SHA256SUM} llvm.src.tar.xz\n${CLANG_SHA256SUM} clang.src.tar.xz" | sha256sum --check -
     - mkdir clang && tar xf clang.src.tar.xz --strip-components=1 --no-same-owner -C clang
     - mkdir llvm && tar xf llvm.src.tar.xz --strip-components=1 --no-same-owner -C llvm
     - mkdir build && cd build
+    # copy llvm cmake modules into the correct directory to ensure discovery
+    # Build runs into this issue if the correct modules cannot be discovered:
+    # https://github.com/llvm/llvm-project/issues/53281
+    - cp ../llvm/Modules/* ../llvm/cmake/modules || echo "LLVM common cmake utils not found in directory 'Modules'"
```
Comment (usamasaqib) [inline .gitlab/deps_build.yml:33]: I am not sure if this is the best way to solve this problem.
Passing a flag to cmake which includes ./Modules directory in the list of include_directories is another option.
We can also patch CMakeLists.txt and explicitly add ./Modules as a directory to search.

**Comment by paulcacheux** [comment]
Hello ! Last I tried, we had some ebpf code that worked with clang-11 but not with clang-12 (some code size limit issue if I remember correctly). Happy to test this deeper when you are ready

**Comment by usamasaqib** [comment]
> Hello ! Last I tried, we had some ebpf code that worked with clang-11 but not with clang-12 (some code size limit issue if I remember correctly). Happy to test this deeper when you are ready

Yes I tested it locally and I saw that some of the programs were not compiling because the stack limits were being exceeded.
The failure happened when assembling the object file with llc.
I wonder if the stack analysis pass done by llc is causing the issue, or did the verifier also complain when you ran your tests?

**Comment by brycekahle** [comment]
@usamasaqib It looks like there is a 14.0.4 now, which you could use. That may solve the `cmake` issue.

**Comment by paulcacheux** [comment]
@usamasaqib https://github.com/DataDog/datadog-agent/pull/12374 should help for the stack size issue

**Pull request closed without merging**